sources/dynamodb/streaming.go

```diff
@@ -253,9 +315,133 @@ func getRecords(streamClient dynamodbstreamsiface.DynamoDBStreamsAPI, shardItera
+	if streamInfo.write == nil {
+		msg := "Internal error: writeRecord called but writer not configured"
+		streamInfo.StatsAddBadRecord(srcTable, eventName)
+		streamInfo.Unexpected(msg)
+	} else {
+		m := getMutation(eventName, srcTable, spTable, spCols, spVals, srcSchema)
+		err := writeMutation(m, streamInfo)
+		if err != nil {
+			streamInfo.StatsAddDroppedRecord(srcTable, eventName)
+			streamInfo.CollectDroppedRecord(eventName, spTable, spCols, spVals, err)
+		}
+	}
+}
+
+// getMutation creates a mutation for writing to Cloud Spanner from the converted data.
+func getMutation(eventName, srcTable, spTable string, spCols []string, spVals []interface{}, srcSchema schema.Table) (m *sp.Mutation) {
+	if eventName == "INSERT" {
+		m = sp.Insert(spTable, spCols, spVals)
+	} else if eventName == "MODIFY" {
+		m = sp.InsertOrUpdate(spTable, spCols, spVals)
+	} else {
+		m = removeMutation(srcSchema, spTable, srcTable, spVals)
+	}
+	return m
+}
+
+// removeMutation create a mutation from converted data for records of type 'REMOVE'.
+// It ensures that when keyset is created the order for primary keys passed is same
+// as the original database i.e. HASH Key, Partition Key.
+func removeMutation(srcSchema schema.Table, spTable, srcTable string, spVals []interface{}) (m *sp.Mutation) {
+	var srcKeys []string
+	var reqSpVals []interface{}
+	for i := 0; i < len(spVals); i++ {
+		if spVals[i] == nil {
+			continue
+		}
+		srcKeys = append(srcKeys, srcSchema.ColNames[i])
+		reqSpVals = append(reqSpVals, spVals[i])
+	}
+	primaryKeys := srcSchema.PrimaryKeys
+	if primaryKeys[0].Column != srcKeys[0] {
+		reqSpVals[0], reqSpVals[1] = reqSpVals[1], reqSpVals[0]
+	}
+	if len(reqSpVals) == 1 {
+		m = sp.Delete(spTable, sp.Key{reqSpVals[0]})
+	} else {
+		m = sp.Delete(spTable, sp.Key{reqSpVals[0], reqSpVals[1]})
+	}
+	return m
+}
+
+// parentDataMissingError checks if the error is a parent data missing error.
+func parentDataMissingError(err error) bool {
+	return strings.Contains(err.Error(), "NotFound") && strings.Contains(err.Error(), "Parent row") && strings.Contains(err.Error(), "is missing")
+}
+
+// writeMutation handles writing of a mutation to Cloud Spanner. If insertion fails
+// because of parent data missing error then it retries for a limit of 1000.
+func writeMutation(m *sp.Mutation, streamInfo *StreamingInfo) error {
+	retryLimit := 1000
+	var err error
+	for retryLimit > 0 {
+		err = streamInfo.write(m)
+		if err == nil || !parentDataMissingError(err) {
+			break
+		}
+		time.Sleep(4 * time.Second)
+		retryLimit--
+	}
+	return err
+}
+
+// setWriter initializes the write function used to write mutations to Cloud Spanner.
+func setWriter(streamInfo *StreamingInfo, client *sp.Client, conv *internal.Conv) {
+	streamInfo.write = func(m *sp.Mutation) error {
+		const migrationMetadataKey = "cloud-spanner-migration-metadata"
```

Comment:
Since this const is used at multiple places it would be better to put it in constants file and use it from there